### PR TITLE
Close drivers on stream initialization failure

### LINKED ISSUE #74: GetUserMedia doesn't close driver if request is partially successed
When calling GetUserMedia with audio and video constraints, if one of audio/video failed to be initialized, other one is left opened. Then, the next GetUserMedia call always fail with `invalid state: driver is already opened`.

### PR DESCRIPTION
**Pull request by at-wat**
In GetUserMedia and GetDisplayMedia, driver was left opened if
initialization failure. Since Trackers are not returned on error,
there was no way to close them.
This commit closes Trackers on GetUserMedia/GetDisplayMedia failure.

### Reference issue
Fixes #74

### CODE REVIEW

File: mediadevices.go [hunk 1]
```diff
@@ -82,6 +82,12 @@ func WithTrackGenerator(gen TrackGenerator) MediaDevicesOption {
 func (m *mediaDevices) GetDisplayMedia(constraints MediaStreamConstraints) (MediaStream, error) {
 	trackers := make([]Tracker, 0)
 
+	cleanTrackers := func() {
+		for _, t := range trackers {
+			t.Stop()
+		}
+	}
+
```
Comment (lherman-cs) [inline mediadevices.go]: I like it!

File: mediadevices_test.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+	"github.com/pion/mediadevices/pkg/io/audio"
+	"github.com/pion/mediadevices/pkg/io/video"
+	"github.com/pion/mediadevices/pkg/prop"
+)
+
+func TestGetUserMedia(t *testing.T) {
+	codec.Register("MockVideo", codec.VideoEncoderBuilder(func(r video.Reader, p prop.Media) (io.ReadCloser, error) {
+		if p.BitRate == 0 {
+			return nil, errors.New("wrong codec parameter")
+		}
```
Comment (lherman-cs) [inline mediadevices_test.go]: Why is this an error?

File: mediadevices_test.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+	}
+
+	// GetUserMedia with correct parameters
+	ms, err = md.GetUserMedia(constraints)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	tracks := ms.GetTracks()
+	if l := len(tracks); l != 2 {
+		t.Fatalf("Number of the tracks is expected to be 1, got %d", l)
```
Comment (lherman-cs) [inline mediadevices_test.go]: Oops, this should be "expected to be 2"

File: mediadevices_test.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+	}
+	tracks[0].OnEnded(func(err error) {
+		if err != io.EOF {
+			t.Errorf("OnEnded called: %v", err)
+		}
+	})
```
Comment (lherman-cs) [inline mediadevices_test.go]: Probably also check for the audio track?

File: mediadevices_test.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+	// Stop and retry GetUserMedia
+	ms, err = md.GetUserMedia(constraints)
+	if err != nil {
+		t.Fatalf("Failed to GetUserMedia after the previsous tracks stopped: %v", err)
+	}
+	tracks = ms.GetTracks()
+	if l := len(tracks); l != 2 {
+		t.Fatalf("Number of the tracks is expected to be 1, got %d", l)
+	}
```
Comment (lherman-cs) [inline mediadevices_test.go]: Should be "expected to be 2" also

File: mediadevices_test.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+	tracks[0].OnEnded(func(err error) {
+		if err != io.EOF {
+			t.Errorf("OnEnded called: %v", err)
+		}
+	})
```
Comment (lherman-cs) [inline mediadevices_test.go]: Probably also check for the audio track

File: track.go [hunk 7]
```diff
@@ -108,6 +108,7 @@ func newVideoTrack(opts *MediaDevicesOptions, d driver.Driver, constraints Media
 	vr := d.(driver.VideoRecorder)
 	r, err := vr.VideoRecord(constraints.Media)
 	if err != nil {
+		_ = d.Close()
```
Comment (lherman-cs) [inline track.go]: I think we should just Close the driver automatically in `driver` package. And, we should simply call the wrapper version of Close so the driver state will be updated too and finally still return the error. It looks to me that this is the `driver` package problem. What do you think?

File: mediadevices_test.go (hunk 2 above)
Comment (at-wat) [reply mediadevices_test.go]: It's just for triggering error during initialization process for testing.

File: track.go (hunk 7 above)
Comment (at-wat) [reply track.go]: Ah, yes. This can be done in the wrapper.

File: mediadevices_test.go (hunk 2 above)
Comment (lherman-cs) [reply mediadevices_test.go]: Ah I see

File: track.go (hunk 7 above)
Comment (at-wat) [reply track.go]: image
So, there will be a choice pseudo-state at the place of `Audio/VideoRecord` and an another one arrow to `Closed` state on failure case.

**Comment by at-wat** [comment]
@lherman-cs addressed your comments!

File: track.go (hunk 7 above)
Comment (lherman-cs) [reply track.go]: Makes sense. Thanks for pointing that out!

**Pull request merged**